wgengine/tsdns/tsdns.go

```diff
@@ -7,128 +7,333 @@
 package tsdns
 
 import (
-	"encoding/binary"
+	"bytes"
+	"context"
 	"errors"
 	"strings"
 	"sync"
 	"time"
 
 	dns "golang.org/x/net/dns/dnsmessage"
 	"inet.af/netaddr"
+	"tailscale.com/net/netns"
 	"tailscale.com/types/logger"
-	"tailscale.com/wgengine/packet"
 )
 
+// maxResponseSize is the maximum size of a response from a Resolver.
+const maxResponseSize = 512
+
+// queueSize is the maximal number of DNS requests that can be pending at a time.
+// If EnqueueRequest is called when this many requests are already pending,
+// the request will be dropped to avoid blocking the caller.
+const queueSize = 8
+
+// delegateTimeout is the maximal amount of time Resolver will wait
+// for upstream nameservers to process a query.
+const delegateTimeout = 5 * time.Second
+
 // defaultTTL is the TTL of all responses from Resolver.
 const defaultTTL = 600 * time.Second
 
+// ErrClosed indicates that the resolver has been closed and readers should exit.
+var ErrClosed = errors.New("closed")
+
 var (
+	errAllFailed      = errors.New("all upstream nameservers failed")
+	errFullQueue      = errors.New("request queue full")
 	errMapNotSet      = errors.New("domain map not set")
-	errNoSuchDomain   = errors.New("domain does not exist")
 	errNotImplemented = errors.New("query type not implemented")
-	errNotOurName     = errors.New("not an *.ipn.dev domain")
-	errNotOurQuery    = errors.New("query not for this resolver")
 	errNotQuery       = errors.New("not a DNS query")
-	errSmallBuffer    = errors.New("response buffer too small")
 )
 
-var (
-	defaultIP   = packet.IP(binary.BigEndian.Uint32([]byte{100, 100, 100, 100}))
-	defaultPort = uint16(53)
-)
-
-// Map is all the data Resolver needs to resolve DNS queries.
+// Map is all the data Resolver needs to resolve DNS queries within the Tailscale network.
 type Map struct {
 	// domainToIP is a mapping of Tailscale domains to their IP addresses.
-	// For example, monitoring.ipn.dev -> 100.64.0.1.
+	// For example, monitoring.tailscale.us -> 100.64.0.1.
 	domainToIP map[string]netaddr.IP
 }
 
 // NewMap returns a new Map with domain to address mapping given by domainToIP.
-// It takes ownership of the provided map.
 func NewMap(domainToIP map[string]netaddr.IP) *Map {
-	return &Map{
-		domainToIP: domainToIP,
-	}
+	return &Map{domainToIP: domainToIP}
 }
 
-// Resolver is a DNS resolver for domain names of the form *.ipn.dev.
+// Packet represents a DNS payload together with the sender address.
+// Since a Resolver always listens on an IPv4 address,
+// uint32 is a suitable (and convenient for interface with wgengine/packet) representation.
+type Packet struct {
+	// Payload is the application layer DNS payload.
+	Payload []byte
+	// SrcIP is the IPv4 address of the packet sender.
+	SrcIP uint32
```

Comment:
I'd prefer to not have uint32s in IPs as IP addresses. Why doesn't netaddr.IP work here?